packages/foam-core/src/model/uri.ts

```diff
@@ -0,0 +1,508 @@
+    }
+  }
+
+  return create('file', authority, path, _empty, _empty);
+};
+
+export const placeholder = (key: string): URI => {
+  return {
+    scheme: 'placeholder',
+    authority: '',
+    path: key,
+    query: '',
+    fragment: '',
+  };
+};
+
+export const from = (
+  uri: URI,
+  change?: {
+    scheme?: string | null;
+    authority?: string | null;
+    path?: string | null;
+    query?: string | null;
+    fragment?: string | null;
+  }
+): URI => {
+  change = change ?? {};
+  const scheme =
+    change.scheme === undefined
+      ? uri.scheme
+      : change.scheme === null
+      ? _empty
+      : change.scheme;
+
+  const authority =
+    change.authority === undefined
+      ? uri.authority
+      : change.authority === null
+      ? _empty
+      : change.authority;
+
+  const path =
+    change.path === undefined
+      ? uri.path
+      : change.path === null
+      ? _empty
+      : change.path;
+
+  const query =
+    change.query === undefined
+      ? uri.query
+      : change.query === null
+      ? _empty
+      : change.query;
+
+  const fragment =
+    change.fragment === undefined
+      ? uri.fragment
+      : change.fragment === null
+      ? _empty
+      : change.fragment;
+
+  return { scheme, authority, path, query, fragment };
+};
+
+export const computeUrisRelativePath = (source: URI, target: URI): string => {
+  const relativePath = posix.relative(posix.dirname(source.path), target.path);
+  return relativePath;
+};
+
+export const getBasename = (uri: URI) => posix.parse(uri.path).name;
+
+export const getDir = (uri: URI) => file(posix.dirname(uri.path));
+
+/**
+ * Uses a placeholder URI, and a reference directory, to generate
+ * the URI of the corresponding resource
+ *
+ * @param placeholderUri the placeholder URI
+ * @param basedir the dir to be used as reference
+ * @returns the target resource URI
+ */
+export const createResourceUriFromPlaceholder = (
+  basedir: URI,
+  placeholderUri: URI
+): URI => {
+  const tokens = placeholderUri.path.split('/');
+  const path = tokens.slice(0, -1);
+  const filename = tokens.slice(-1);
+  return joinPath(basedir, ...path, `${filename}.md`);
+};
+
+/**
+ * Join a URI path with path fragments and normalizes the resulting path.
+ *
+ * @param uri The input URI.
+ * @param pathFragment The path fragment to add to the URI path.
+ * @returns The resulting URI.
+ */
+export const joinPath = (uri: URI, ...pathFragment: string[]): URI => {
+  if (!uri.path) {
+    throw new Error(`[UriError]: cannot call joinPath on URI without path`);
+  }
+  let newPath: string;
+  if (isWindows && uri.scheme === 'file') {
+    newPath = file(paths.win32.join(toFsPath(uri), ...pathFragment)).path;
+  } else {
+    newPath = paths.posix.join(uri.path, ...pathFragment);
+  }
+  return from(uri, { path: newPath });
+};
+
+export const uriToSlug = (uri: URI): string => {
+  return GithubSlugger.slug(posix.parse(uri.path).name);
+};
+
+export const uriToHash = (uri: URI): string => {
```

Comment:
Is this used? If it is, should live in its own module and not on URI. If not, nix?